### PR TITLE
Backport 2.28: Report configuration settings in the outcome file

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Backport of https://github.com/Mbed-TLS/mbedtls/pull/9172 and its framework companion https://github.com/Mbed-TLS/mbedtls-framework/pull/28.

I followed the same commit structure for the common content, but several things are different in 2.28.

* Minor fixes from the main PR.
* Test code and manually written test cases.
* Commits from the framework PR to create the Python script at a temporary location.
* Move the script to its pre-framework location.
* Adapt the script content for 2.28. Maybe `dependencies_of_setting` needs more 2.28-only stuff but we can do that later based on looking at outcome analysis.
* Check in the generated files, since we do that in 2.28.
* Add the generated files to `check-generated-files`. In 2.28, there is no handling of generated files in `**/Makefile` and `**/CMakeLists.txt`, and no framework submodule to update.

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** no (test only)
- [x] backport of https://github.com/Mbed-TLS/mbedtls/pull/9172
- [x] **tests** provided

### CODE REVIEW

**Review by ronald-cron-arm — COMMENTED** [review]
Just to be sure, do we really need that in 2.28? Its end of life is at the end of the year thus what the point to "Check that all configuration options are tested in all.sh" now for it?

**Comment by gilles-peskine-arm** [comment]
I don't think it's worth it to _enforce_ the same level of configuration testing in 2.28. And in that respect, I didn't go digging for options with dependencies in 2.28 (options where A requires B so we want to check A:B and !A:B, not just A and !A), not even to the small level I did for 3.6. But to even evaluate this, I need to get the reports from this pull request. Doing the backport was only about ½h of engineering time, plus the time to review it. I think that's well worth it.

File: tests/scripts/check-generated-files.sh [hunk 1]
```diff
@@ -102,4 +102,5 @@ check scripts/generate_visualc_files.pl visualc/VS2010
 check scripts/generate_psa_constants.py programs/psa/psa_constant_names_generated.c
 check tests/scripts/generate_psa_wrappers.py tests/include/test/psa_test_wrappers.h tests/src/psa_test_wrappers.c
 check tests/scripts/generate_bignum_tests.py $(tests/scripts/generate_bignum_tests.py --list)
+check tests/scripts/generate_config_tests.py $(tests/scripts/generate_bignum_tests.py --list)
```
Comment (paul-elliott-arm) [inline tests/scripts/check-generated-files.sh:105]: Was this deliberate or copypasta? I don't think `generate_config_tests.py` has a `--list` option, but is the `generate_bignum_tests.py` list compatible?

**Review by paul-elliott-arm — COMMENTED** [review]
LGTM, just one question.

File: tests/scripts/check-generated-files.sh (hunk 1 above)
Comment (gilles-peskine-arm) [reply tests/scripts/check-generated-files.sh:105]: Woops, copypasta. Fixed.

All the `generate_xxx_tests.py` use [the same `main` function from `test_data_generation.py`](https://github.com/Mbed-TLS/mbedtls-framework/blob/main/scripts/mbedtls_framework/test_data_generation.py#L179) and they take the same command line options.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
I've checked the backport and reviewed the additional commits. One comment otherwise this looks good to me.

**Review by ronald-cron-arm — APPROVED** [review]
LGTM, thanks.

**Review by paul-elliott-arm — APPROVED** [review]
LGTM

**Pull request merged**